Support storage and buckets for self hosted

adityadhawan22 opened the pull request:
## What kind of change does this PR introduce?
Enable storage/bucket for self-hosted supabase

https://user-images.githubusercontent.com/15171755/186464204-6a548d95-f18d-4799-9b28-db33abaaf748.mov


Closes #4440

Copystrike commented:
Nice PR! Only I found an issue running this with docker.

# The issue

When opening the storage page in supabase studio, it will try to fetch the buckets. The issue is that its trying to fetch the buckets from the wrong URL:
![image](https://user-images.githubusercontent.com/26123873/187051720-cdf57f94-19be-486d-8292-70e322016f83.png)


# Why this is happening

Looking into the codebase, the way the supabase client knows to what URL it has to connect it will make a request to 
`http://localhost:3000/api/props/project/default/settings` and it will get the value from `app_config.endpoint` 

[Link to source code](https://github.com/supabase/supabase/blob/43f85cceadb0193a274c9e7ef404a0e83c25fd66/studio/components/layouts/StorageLayout/StorageLayout.tsx#L49)
![image](https://user-images.githubusercontent.com/26123873/187051746-a7d2df97-be63-4b01-9761-5567b82853da.png)

And by looking at the code from that page it turns out that its get the url from `process.env.SUPABASE_URL`

[Link to source code](https://github.com/supabase/supabase/blob/0ce37f10eb3c8b18f49383925e983c9c3ee06af9/studio/pages/api/props/project/%5Bref%5D/settings.ts#L58)
![image](https://user-images.githubusercontent.com/26123873/187051516-11749d04-8ab1-4827-b816-16fea5ac5886.png)

Which by default is set to `http://kong:8000` for the official docker compose project (on purpose?)

[Link to docker-compose file](https://github.com/supabase/supabase/blob/1df5be4f4e00ae8bdb3083059ac1503127556bac/docker/docker-compose.yml#L20)
![image](https://user-images.githubusercontent.com/26123873/187051765-63e3298d-5fde-4e55-9f84-eb2d6ac018ac.png)

# Solutions

These are solutions I think will work, but I have no idea if they are secure since I have no idea how other features use the superbase client.

1. A new environment variable that exposes kong url, so `http://localhost:3000/api/props/project/default/settings` exposes the right url address and not the docker URL.
2. Change `SUPABASE_URL` to `http://localhost:8000` in the docker-compose.yml.

# Docker

I made an docker image of this PR and merged it with the latest commit and pushed it to docker hub for anyone that wanna test this [Docker hub image](https://hub.docker.com/repository/docker/localplayer/studio)
(Don't forget to change the studio image name in docker-compose.yml)

adityadhawan22 commented:
> When opening the storage page in supabase studio, it will try to fetch the buckets. The issue is that its trying to fetch the buckets from the wrong URL: 



Good find. I guess changing the `SUPABASE_URL` to  `http://localhost:8000` would be an easy fix. An alternative solution would be that I can create a proxy for storage API that calls `http://kong:8000` on the next server. Would be nice if someone from supabase team can review my PR

kiwicopple commented:
I think we had the same issue with the API docs, and that's when we added the `PUBLIC_REST_URL` variable:
https://github.com/supabase/supabase/blob/5b16471c7a84f17957c0f959353048aac8b30faa/docker/.env.example#L74

I think that the most ergonomic solution is:

1. rename `PUBLIC_REST_URL` to `SUPABASE_PUBLIC_URL`  (with the value `http://localhost:8000`)
2. replace anywhere that `PUBLIC_REST_URL` is used, and in these cases it will need to append the `/rest/v1` path
3. update the storage section to use the new `SUPABASE_PUBLIC_URL`, and in this case it will need to append the `/storage/v1` path

Does that make sense? I'm just skimming the code - so feel free to push back if that seems like it's not the correct approach

Copystrike commented:
> I think we had the same issue with the API docs, and that's when we added the `PUBLIC_REST_URL` variable:
> 
> https://github.com/supabase/supabase/blob/5b16471c7a84f17957c0f959353048aac8b30faa/docker/.env.example#L74
> 
> I think that the most ergonomic solution is:
> 
> 1. rename `PUBLIC_REST_URL` to `SUPABASE_PUBLIC_URL`  (with the value `http://localhost:8000`)
> 2. replace anywhere that `PUBLIC_REST_URL` is used, and in these cases it will need to append the `/rest/v1` path
> 3. update the storage section to use the new `SUPABASE_PUBLIC_URL`, and in this case it will need to append the `/storage/v1` path
> 
> Does that make sense? I'm just skimming the code - so feel free to push back if that seems like it's not the correct approach

Sounds like a good idea! This will also be good for later on if there needs to be more API endpoints so we dont have to add more environment variables for every endpoint.

Looking through the .env file I saw there is acually already an environment variable called "API_EXTERNAL_URL" which already has the good URL, but it is in the section of "auth", maybe we could make that a global environment variable instead ? Unless we want to keep things seperated.

https://github.com/supabase/supabase/blob/5b16471c7a84f17957c0f959353048aac8b30faa/docker/.env.example#L46

Anyways, the solution is pretty good and if I find some free time this week I'll be able to test it out and make a PR into this project if wanted.

Commit:
Support storage and buckets for self hosted

Commit:
Update .env.example

adityadhawan22 commented:
> I think we had the same issue with the API docs, and that's when we added the `PUBLIC_REST_URL` variable:
> 
> https://github.com/supabase/supabase/blob/5b16471c7a84f17957c0f959353048aac8b30faa/docker/.env.example#L74
> 
> I think that the most ergonomic solution is:
> 
> 1. rename `PUBLIC_REST_URL` to `SUPABASE_PUBLIC_URL`  (with the value `http://localhost:8000`)
> 2. replace anywhere that `PUBLIC_REST_URL` is used, and in these cases it will need to append the `/rest/v1` path
> 3. update the storage section to use the new `SUPABASE_PUBLIC_URL`, and in this case it will need to append the `/storage/v1` path
> 
> Does that make sense? I'm just skimming the code - so feel free to push back if that seems like it's not the correct approach

Cool, I have changed the `PUBLIC_REST_URL` to `SUPABASE_PUBLIC_URL` 
Can you verify the PR?  😃

jenspots commented:
Any updates on this? I'd love to have this feature on self hosted instances!

kiwicopple commented:
Will get the team to resolve these conflicts and check 👍

pucho commented:
@adityadhawan22 Have you had this issue before? I'm working on this today, just to make sure it's not only on my side. Thanks!
<img width="1432" alt="Screen Shot 2022-11-15 at 13 09 09" src="https://user-images.githubusercontent.com/1002620/201969060-0232b5df-1c0c-4468-8421-3085664c61d3.png">

pucho committed:
Merge branch 'master' into feat/self-hosted-storage

pucho commented:
@adityadhawan22 @jenspots I resolved some conflicts with our latest changes.

adityadhawan22 commented:
> this issue

yes, I was having this error too, it seems like a macOS only error

pucho commented:
> > this issue
> 
> yes, I was having this error too, it seems like a macOS only error
> 
> I assume you are running it on macos+docker too?

Yes, same setup here.

adityadhawan22 commented:
> > > this issue
> > 
> > 
> > yes, I was having this error too, it seems like a macOS only error
> > I assume you are running it on macos+docker too?
> 
> Yes, same setup here.

https://github.com/supabase/storage-api/issues/201

sweatybridge commented on studio/components/layouts/StorageLayout/StorageMenu.tsx:
```diff
@@ -18,6 +18,7 @@ import {
 import ProductMenuItem from 'components/ui/ProductMenu/ProductMenuItem'
 import { STORAGE_ROW_STATUS } from 'components/to-be-cleaned/Storage/Storage.constants'
 import { useStorageStore } from 'localStores/storageExplorer/StorageExplorerStore'
+import { IS_PLATFORM } from 'lib/constants'
```
Seems unused

sweatybridge commented on studio/pages/api/props/project/[ref]/api.ts:
```diff
@@ -72,7 +72,7 @@ const handleGetAll = async (req: NextApiRequest, res: NextApiResponse) => {
       },
       endpoint: process.env.SUPABASE_URL,
-      restUrl: process.env.SUPABASE_REST_URL,
+      restUrl: `${process.env.SUPABASE_REST_URL}/rest/v1/`,
```
Can we revert this change? The original problem with misconfigured storage route has already been addressed. So this might not be relevant anymore. 

In addition, changing restUrl also requires updating the cli to change its env var when updating to newer studio image. Technically a braking change.

sweatybridge commented on docker/.env.example:
```diff
@@ -74,4 +74,4 @@ STUDIO_DEFAULT_ORGANIZATION=Default Organization
 
 STUDIO_PORT=3000
-PUBLIC_REST_URL=http://localhost:8000/rest/v1/ # replace if you intend to use Studio outside of localhost
+SUPABASE_PUBLIC_URL=http://localhost:8000 # replace if you intend to use Studio outside of localhost
```
```suggestion
SUPABASE_PUBLIC_URL=http://localhost:8000/rest/v1/ # replace if you intend to use Studio outside of localhost
```

sweatybridge committed:
Merge branch 'master' into feat/self-hosted-storage

sweatybridge committed:
chore: use supabase public url

sweatybridge commented on docker/docker-compose.yml:
```diff
@@ -20,8 +20,8 @@ services:
-      SUPABASE_URL: http://kong:8000
-      SUPABASE_REST_URL: ${PUBLIC_REST_URL}
+      SUPABASE_URL: ${SUPABASE_URL}
+      SUPABASE_REST_URL: ${SUPABASE_URL}/rest/v1/
```
kept for backwards compatibility with `studio:0.22.08`

sweatybridge commented on studio/components/layouts/ProjectLayout/NavigationBar/NavigationBar.utils.tsx:
```diff
@@ -84,23 +84,18 @@ export const generateProductRoutes = (ref?: string, project?: ProjectBase): Rout
+          ? homeUrl
+          : isProjectBuilding
+          ? buildingUrl
+          : `/project/${ref}/storage/buckets`),
```
TODO: add this route to `api/pages`

sweatybridge committed:
fix: docker release workflow

sweatybridge committed:
Merge branch 'master' into feat/self-hosted-storage

sweatybridge committed:
chore: update gotrue variables

Pranavprana commented:
I used the latest version still no storage GUI can you guide me what wrong am doing

masoudei commented:
can I help you fixing this issue and let the world use all features in UI ?

dominikjannis commented:
any updates to this?

sweatybridge commented:
@dominikjannis Storage UI is now available in self-hosted :)
